posts/2023-12-19-dt_particularities-toby_hocking/index.qmd

```diff
@@ -0,0 +1,152 @@
+---
+title: "Particular syntax and semantics of {data.table}"
+author: "Toby Hocking"
+date: "2023-12-19"
+categories: [tutorials]
+draft: false
+---
+
+The goal of this blog post is to explain some similarities and
+differences between base R data frames, and data tables. We will
+focus on accessing/assigning values, and discuss two major
+differences:
+
+* Syntax means the structure of the code that is used. `data.table`
+  uses a syntax where most operations can be done within the square
+  brackets: `DT[i,j,by]`.
+* Semantics means how the data structure/variable works. We say that a
+  `data.table` has reference semantics, meaning we can modify a
+  `data.table` from within a function, and see those modifications
+  after the function is done executing. In other words, two different
+  R variables can point to, and modify, the same `data.table`.
+
+## Difference in syntax
+
+To discuss the similarities and differences in syntax, consider the data below,
+
+```{r}
+library(data.table)
+syntax <- function(type, name, columns, code){
+  mcall <- match.call()
+  dt.args <- lapply(as.list(mcall[-1]), paste)
+  do.call(data.table, dt.args)
+}
+(syntax.tab <- rbind(
+  syntax(frame, literal, one, "DF$col <- value"),
+  syntax(table, literal, one, "DT[, col := value]"),
+  syntax(frame, variable, multiple, 'DF[,cols] <- values'),
+  syntax(table, variable, multiple, 'DT[,(cols) := values]')))
+```
+
+The table above defines the different syntax required to do column
+assignment in data tables and frames. 
+
+* `type` indicates data type: frame or table.
+* `name` indicates whether the column name to assign is literally
+  written in the code, or if a variable is specified (and the column
+  name is taken from the value of that variable).
+* `columns` indicates whether only one or multiple (one or more) columns
+  can be assigned using the syntax.
+* `code` is the syntax of the R code used for the assignment. Note
+  that `col` is a literal column name, whereas `cols` is a variable
+  (vector of column names or numbers).
+* Note that there are other ways to do column assignment. For example,
+  `DF[["col"]]<-value` can also be used for single column assignment
+  in a data frame. And `set(DT,j=cols,value=values)` is a more
+  efficient version of column assignment for data tables (no overhead
+  of `[.data.table` method), that is recommended for use in loops.
+
+Below we reshape those data to
+facilitate comparison between frame and table versions:
+
+```{r}
+options(width=100)
+data.table::dcast(syntax.tab, name + columns ~ type, value.var="code")  
```

Comment:
I think we should skip `data.table::` prefix and assume pkg to be attached

Reply:
typically I do skip data.table:: in my own code, but when I am teaching, I prefer to be explicit about what package each object comes from (especially in this case where there is another dcast in reshape2)